docs/adr/2023-08-18_025-resource-based-api.md

```diff
@@ -0,0 +1,146 @@
+
+  - As [ADR-15](/adr/15) also proposes, some clients may not need (or should
+    not have) access to administrative information.
+
+- It is often a good idea to separate the responsibilities of Commands and
+  Queries (CQRS), as well as the model they use.
+
+## Decision
+
+- Do not add a general purpose querying interface (e.g. graphql) to the
+  `hydra-node` and stay true to [ADR-3](/adr/3) of using duplex communication
+  channels (using websockets).
+
+- **Drop** `GetUTxO` and `GetUTxOResponse` messages as they advocate a
+  request/response way of querying.
+
+- Realize that `ClientInput` data is actually a `ClientCommand` (renaming them)
+  and that `ServerOutput` are just `projections` of the [internal event stream
+  (see ADR-24)](/adr/24) into read `models` on the API layer.
+
+- Compose the API out of resource `models`, which compartmentalize the domain
+  into topics on the API layer.
+
+  - Each resource has a _latest_ state, which can be queried or clients may
+    subscribe to any model changes.
+
+  - A resource's `model` type needs to be a result of a pure `projection` from
+    server output events, i.e. `project :: model -> ServerOutput -> model`.
+
+    - TBD: or use the event stream type `StateChanged` instead? potentially
+      renaming it?
+
+  - Each resource is available at some HTTP path, also called "endpoint":
+
+    - `GET` requests must respond with the _latest_ state in a single response.
+
+    - `GET` requests with websocket upgrade headers must start a websocket
+      connection, push the _latest_ state as first message and any resource
+      state updates after.
+
+    - Other HTTP verbs may be accepted by a resource handler, i.e. to issue
+      resource-specific commands. Any commands accepted must also be available
+      via the corresponding websocket connection.
+
+  - Each resource is only available as `JSON` encoding (for now).
+
+    - (TBD): already specify that `Accept` headers should be used to negotiate a
+      `content-type`?
+
+- (TBD): Keep the direct/raw output of `ServerOutput` events on `/`, which
+  also accepts all `ClientCommand` messages.
+
+### Resources
+
+Resource paths + HTTP verbs mapped to existing things:
+
+| Path                        | GET                         | POST                   | PATCH   | DELETE             |
+| :-------------------------- | :-------------------------- | :--------------------- | ------- | :----------------- |
+| `/node/head/status`         | `HeadStatus(..)`            | -                      | -       | -                  |
+| `/node/head/utxo`           | confirmed utxo              | -                      | -       | -                  |
+| `/node/head/transactions`   | confirmed txs               | `NewTx` + responses    | -       | -                  |
+| `/node/head/ledger`         | `localUTxO` and `localTxs`  | -                      | -       | -                  |
```

Comment:
Now I think I understand the following:

* `/node/head/utxo` -> utxo set of the last known validated snapshot
* `/node/head/transactions` -> transactions included in the last known validated snapshot (transitively)
* `/node/head/ledger`  -> utxo set and transactions of the local ledger, irrespective of the snapshots

I'm not sure we want to expose so much details to the customer. This notion of local transactions and local utxo set feels very internal to the Hydra protocol specification/implementation.

I think, as a client of Hydra, that I want to be able to :
1. init/open/close/contest/fanout a head
2. build new L2 transactions
3. know about the finality of a transactions on L2

For 1. I need the most recent view of the ledger from my node. Otherwise, I could build locally invalid transactions (see #612 ) so `/node/head/ledger` would be enough, although I would name it differently.

For 2. I need the information that a given transaction has been included in a signed snapshot and, for that, I guess I would want to have an entry point to get the status of the transaction itself and it would inform me about the fact that it's _pending_ and _snapshotted_.

That would lead to the following entry points:

* `/node/head/utxo` -> utxo set known by my local node
* `/node/head/transactions` -> transactions known by my local node
* * `/node/head/transactions/<id>` -> informations about transaction of id _id_, can be used to check for L2 finality of a transaction
* `/node/head/ledger`  -> removed

Reply:
I would say exactly the other way around?

For 1) you would care about which UTxO are available on the L2 after opening and which would be available if we close (any non confirmed utxo would not be included!)

For 2) you would need the most recent local view of the UTxO set, such that your node accepts transactions moving this forward.

For 3) we can do this individual query as you propose, but I would advise against it as it does not allow us to setup a subscription for all confirmed L2 transactions, which is a use case we know exists.As you might 

So we are in the situation where likely need both, the local view and the confirmed view. I will try to make that separation more clear in the example resources and also should make it clear that the ADR is not primarily about specifying the full API, but more to exercise the concept of mapping endpoints to existing semantics and see whether we could address the requirements from the context.